substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrThrottlerWindow.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2023, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, 2023, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core.jfr;
+
+import com.oracle.svm.core.jdk.UninterruptibleUtils;
+
+import static java.lang.Math.log;
+
+public class JfrThrottlerWindow {
+    // reset every rotation
+    public UninterruptibleUtils.AtomicLong measuredPopSize;
+    public UninterruptibleUtils.AtomicLong endTicks;
+
+    // Calculated every rotation based on params set by user and results from previous windows
+    public volatile long samplingInterval;
+    public volatile double maxSampleablePopulation;
+
+    // Set by user
+    public long samplesPerWindow;
+    public long windowDurationNs;
+    public long debt;
+
+    public JfrThrottlerWindow() {
+        windowDurationNs = 0;
+        samplesPerWindow = 0;
+        maxSampleablePopulation = 0;
+        measuredPopSize = new UninterruptibleUtils.AtomicLong(0);
+        endTicks = new UninterruptibleUtils.AtomicLong(JfrTicks.currentTimeNanos() + windowDurationNs);
+        samplingInterval = 1;
+        debt = 0;
+    }
+
+    /**
+     * A rotation of the active window could happen while in this method. If so, then this window
+     * will be updated as usual, although it is now the "next" window. This results in some wasted
+     * effort, but doesn't affect correctness because this window will be reset before it becomes
```

Comment:
It doesn't seem like hotspot does anything to prevent such a double rotation. If a double rotation happens while a thread is busy sampling the following could happen:
1. The sample gets counted towards the new window (increases the measured population). It could be argued that this is ok since the evaluation happened during the new window's time slice. Depends on when we define a sample as being "taken".
2. Sampling races with the rotating thread writing new window parameters. This can cause it to either sample when it would otherwise not have, or not sample when it otherwise would have. Under sampling should be fine in this edge case. Oversampling is bad.

To avoid those problems and any ambiguity about which window a sample belongs to, I've added a read-write lock. The writer's lock must be acquired for window rotation. The reader's lock must be acquired to sample. 

There is locking without transition so I've made the appropriate code `uninterruptible`. This also required avoiding `Math.Random` and instead copying the simple PRNG that JFR uses in hotspot.